### PR TITLE
Fix amp issues with using length argument for softmax in attention

### PR DESCRIPTION
**Pull request by fhieber**
Re-introduce (again and again) the use of the length argument in the softmax call in DotAttentionCell.
Previously this caused issues with Automatic Mixed Precision training as broadcast_like does not like different dtypes for its input arguments. While this has been fixed in MXNet master, it was not yet merged back to the 1.x branches, therefore it is not part of MXNet 1.6 or 1.7 (or 1.8).

However, we can move the creation of the length tensor for softmax to a different place which avoids the differing dtype problem. 

This change is confirmed to be working with AMP and giving a slight increase in training and inference speed as previously measured.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/decoder.py [hunk 1]
```diff
@@ -286,23 +282,24 @@ def forward(self, step_input, states):
     def hybrid_forward(self, F, step_input, states):
         mask = None
         if self.inference_only:
-            steps, source_mask, *other = states
+            steps, att_valid_length, *other = states
```
Comment (ketranm) [inline sockeye/decoder.py:285]: Is `att_valid_length` the valid length of source sequence. Maybe we can made it more explicit (e.g., `source_att_valid_length`)

File: sockeye/decoder.py [hunk 2]
```diff
@@ -286,23 +282,24 @@ def forward(self, step_input, states):
...
 
         if any(layer.num_state_tensors > 1 for layer in self.layers):
             # separates autoregressive states by layer
             states_iter = iter(autoregr_states)
             autoregr_states = [list(islice(states_iter, 0, layer.num_state_tensors)) for layer in self.layers]
 
-        # Fold the heads of source_mask (batch_size, num_heads, seq_len) -> (batch_size * num_heads, 1, seq_len)
-        source_mask = F.expand_dims(F.reshape(source_mask, shape=(-3, -2)), axis=1)
+        # (batch_size * heads, query_length)
+        att_valid_length = layers.prepare_softmax_lengths(F, att_valid_length, step_input,
```
Comment (ketranm) [inline sockeye/decoder.py:301]: I can guess what this function does, but its name is a bit ambiguous.

File: sockeye/decoder.py (hunk 1 above)
Comment (fhieber) [reply sockeye/decoder.py:285]: good point, renamed

File: sockeye/decoder.py (hunk 2 above)
Comment (fhieber) [reply sockeye/decoder.py:301]: Do you have a better suggestion? I can't think of a better alternative.

Comment (ketranm) [reply sockeye/decoder.py]: I can only think of a boring name `prepare_source_valid_lengths` :)

Comment (fhieber) [reply sockeye/decoder.py]: ok will do

**Review by ketranm — APPROVED** [review]

**Pull request merged**